### PR TITLE
PMM-8269 MongoDB Point-in-Time backups

### PR DESCRIPTION
**Pull request by artemgavrilov**
[PMM-8269](https://jira.percona.com/browse/PMM-8269)

- [ ] Build: https://github.com/Percona-Lab/pmm-submodules/pull/2001
---
- [ ] https://github.com/percona/pmm/pull/769
- [ ] https://github.com/percona/pmm-managed/pull/811
---

### CODE REVIEW

File: actions/concurrent_runner.go [hunk 1]
```diff
@@ -74,6 +78,24 @@ func (r *ConcurrentRunner) Start(a Action, timeout time.Duration) {
 		r.l.Errorf("Ignoring Start: %s.", err)
 		return
 	}
+	actionID, actionType := a.ID(), a.Type()
+
+	var requirements []lock.Entity
+	switch actionType {
+	case "pbm-switch-pitr":
+		requirements = append(requirements, lock.PBM)
+	}
```
Comment (artemgavrilov) [inline actions/concurrent_runner.go:87]: I suppose this should be moved tho the actions/jobs, so they will be able to provide their requirements. But I prefer to do that in a separate ticket.

File: jobs/common.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+func (c *PBMConfig) Equals(config *PBMConfig) bool {
+	if c.PITR.Enabled != config.PITR.Enabled {
+		return false
+	}
+
+	if c.Storage.Type != config.Storage.Type {
+		return false
+	}
+
+	if c.Storage.S3.Bucket != config.Storage.S3.Bucket {
```
Comment (Dasio) [inline jobs/common.go:60]: these checks should be doen only if storage.type == S3, right?

File: client/client.go [hunk 3]
```diff
@@ -37,6 +37,8 @@ import (
 	"google.golang.org/grpc/credentials"
 	grpcstatus "google.golang.org/grpc/status"
 
+	"github.com/percona/pmm-agent/lock"
+
```
Comment (Dasio) [inline client/client.go:41]: empty line

File: jobs/common.go (hunk 2 above)
Comment (artemgavrilov) [reply jobs/common.go:60]: All sub-structs are added as values (not pointers) to the parent struct, so we can just go through every field and check equality without nil checks.

File: client/client.go (hunk 3 above)
Comment (artemgavrilov) [reply client/client.go]: Fixed

**Review by Dasio — APPROVED** [review]

File: actions/concurrent_runner.go [hunk 4]
```diff
@@ -74,6 +78,24 @@ func (r *ConcurrentRunner) Start(a Action, timeout time.Duration) {
 		r.l.Errorf("Ignoring Start: %s.", err)
 		return
 	}
+	actionID, actionType := a.ID(), a.Type()
+
+	var requirements []lock.Entity
+	switch actionType {
+	case "pbm-switch-pitr":
```
Comment (oter) [inline actions/concurrent_runner.go:85]: let's try to improve this switch statement to not to use string literal, e.g. define some string constant in the actions package and use it here and in the `Type() string` function.

**Review by oter — APPROVED** [review]

File: jobs/runner.go [hunk 5]
```diff
@@ -60,6 +64,30 @@ func (r *Runner) Run(ctx context.Context) {
...
+
+			var requirements []lock.Entity
+			switch jobType {
+			case MongoDBBackup, MongoDBRestore:
+				requirements = append(requirements, lock.PBM)
+			case MySQLBackup, MySQLRestore:
+				// TODO: add requirements
+			}
+
+			if !r.locksService.TryAcquire(requirements...) {
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply jobs/runner.go:77]: +1

File: actions/concurrent_runner.go (hunk 4 above)
Comment (artemgavrilov) [reply actions/concurrent_runner.go:85]: https://github.com/percona/pmm-agent/pull/285#discussion_r701123746

I tried to keep the scope of this PR small. Will do that.

File: lock/lock.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+// New creates new locks service.
+func New() *Service {
+	return &Service{
+		entities: make(map[Entity]struct{}),
+	}
+}
+
+// TryAcquire checks that passed entities can be locked. If all passed entities available for locking
+// returns ture, otherwise returns false. This method locks all or nothing.
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply lock/lock.go]: Done

File: actions/concurrent_runner.go (hunk 4 above)
Comment (artemgavrilov) [reply actions/concurrent_runner.go]: Removed at all

File: jobs/runner.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply jobs/runner.go]: Removed at all

File: jobs/common.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+// PBMConfig represents pbm configuration file.
+type PBMConfig struct {
+	Storage Storage `yaml:"storage"`
+	PITR    PITR    `yaml:"pitr"`
+}
+
+// Equals compares two PBMConfig's.
+func (c *PBMConfig) Equals(config *PBMConfig) bool {
+	if c.PITR.Enabled != config.PITR.Enabled {
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply jobs/common.go]: Removed at all

**Review by oter — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**